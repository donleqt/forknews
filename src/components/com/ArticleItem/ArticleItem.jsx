import React from 'react';
import {config} from '../../../config/config';
import {dateTimeToDateRender} from '../../../helpers/utils/time';
import PropsTypes from 'prop-types';

export class ArticleItem extends React.PureComponent {
  static propsTypes = {
    data: PropsTypes.object.isRequired,
    onArticleClick: PropsTypes.func
  };

  onArticleClick = () => {
    if (this.props.onArticleClick) {
      this.props.onArticleClick();
    }
  };

  render() {
    const {snippet, multimedia, pub_date, source} = this.props.data;
    return (
        <div className="article-item" onClick={this.onArticleClick}>
          {!!multimedia && (
              <div className="article-thumbnail">
                <img src={`${config.urlAssets}${multimedia}`} className="article-img"/>
              </div>
          )}
          <div className="article-content">
            <div className="snippet">{snippet}</div>
            {!!source && <div className="source">{source}</div>}
            {!!pub_date && <div className="date">{dateTimeToDateRender(pub_date)}</div>}
          </div>
        </div>
    );
  }
}