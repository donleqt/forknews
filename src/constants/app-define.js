// Status type
export const StatusTypes = {
  loading: 'loading',
  success: 'success',
  error: 'error'
};