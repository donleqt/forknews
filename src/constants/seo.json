{
  "title": "Fork News - See the world moving",
  "description": "See the world moving - powered by NY Times"
}