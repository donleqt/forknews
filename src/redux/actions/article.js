import {ActionTypes} from '../action-types';
import {NetService} from '../../services/net/net';
import {config} from '../../config/config';

// Create net for api call, ge
const {net} = new NetService({
  getResponse: resp => resp && resp.response,
  query: {
    'api-key': config.apiKey
  }
});

export const articleActions = {
  fetchList: ({q, page}) => async dispatch => {
    dispatch({
      type: ActionTypes.FETCH_ARTICLE_LIST
    });
    const resp = await net.get(config.api.articleSearch, {
      query: {
        page: page - 1,
        q: q
      }
    });
    if (resp) {
      dispatch({
        type: ActionTypes.FETCH_ARTICLE_LIST_SUCCESS,
        data: resp.docs,
        meta: resp.meta
      });
    }
    else {
      dispatch({
        type: ActionTypes.FETCH_ARTICLE_LIST_FAIL
      });
    }
    return resp;
  },
  viewArticle: (article) => {
    return {
      type: ActionTypes.VIEW_ARTICLE,
      data: article
    }
  },
  exitArticle: () => {
    return {
      type: ActionTypes.EXIT_ARTICLE
    }
  }
};