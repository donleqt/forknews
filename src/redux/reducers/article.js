import {ActionTypes} from '../action-types';
import {StatusTypes} from '../../constants/app-define';

export const initialState = {
  list: {
    status: null,
    data: null,
    meta: null,
  },
  detail: null
};

// Parse data to standard for UI
export const parseData = data => {
  return {
    id: data._id,
    snippet: data.snippet,
    multimedia: data.multimedia[0] && data.multimedia[0].url,
    pub_date: data.pub_date && new Date(data.pub_date),
    source: data.source,
  };
};

// Calculate standard meta data
export function calculateMeta(meta, pageSize) {
  const totalPage = Math.ceil(meta.hits / pageSize);
  return {
    total: meta.hits,
    totalPage: Math.min(totalPage, 201),
    page: Math.ceil(meta.offset / pageSize) + 1
  }
}

export default function (state = initialState, action) {
  const newState = {...state};
  switch (action.type) {
    case ActionTypes.FETCH_ARTICLE_LIST: {
      newState.list = {
        ...newState.list,
        status: StatusTypes.loading
      };
      break;
    }
    case ActionTypes.FETCH_ARTICLE_LIST_SUCCESS: {
      newState.list = {
        status: StatusTypes.success,
        data: action.data.map(parseData),
        meta: calculateMeta(action.meta, action.data.length)
      };
      break;
    }
    case ActionTypes.FETCH_ARTICLE_LIST_FAIL: {
      newState.list = {
        status: StatusTypes.error,
        data: null,
        meta: action.meta
      };
      break;
    }
    case ActionTypes.VIEW_ARTICLE: {
      newState.detail = action.data;
      break;
    }
    case ActionTypes.EXIT_ARTICLE: {
      newState.detail = null;
      break;
    }
  }
  return newState;
}