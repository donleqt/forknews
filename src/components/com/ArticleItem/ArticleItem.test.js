import renderer from 'react-test-renderer';
import {articles} from 'constants/dummy';
import {parseData} from 'redux/reducers/article';
import {ArticleItem} from './ArticleItem';
import React from 'react';

describe('Test ArticleItem', () => {
  it('Should render article with right snapshot', () => {
    const article0 = articles[0];
    const component = renderer.create(
        <ArticleItem data={parseData(article0)}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});