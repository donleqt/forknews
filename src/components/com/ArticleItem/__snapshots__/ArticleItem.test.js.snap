// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ArticleItem Should render article with right snapshot 1`] = `
<div
  className="article-item"
  onClick={[Function]}
>
  <div
    className="article-thumbnail"
  >
    <img
      className="article-img"
      src="https://static01.nyt.com/images/2018/02/25/travel/25bites/25bites-articleLarge.jpg"
    />
  </div>
  <div
    className="article-content"
  >
    <div
      className="snippet"
    >
      At his new restaurant Folklore, the chef Damian D’Silva continues to celebrate so-called heritage food — and is perhaps starting a few traditions of his own.
    </div>
    <div
      className="source"
    >
      The New York Times
    </div>
    <div
      className="date"
    >
      February 24
    </div>
  </div>
</div>
`;
